cmd/data-usage.go

```diff
@@ -0,0 +1,140 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/pkg/hash"
+)
+
+const (
+	dataUsageObjName       = "data-usage"
+	dataUsageCrawlInterval = 12 * time.Hour
+)
+
+func initDataUsageStats() {
+	go runDataUsageInfoUpdateRoutine()
+}
+
+func runDataUsageInfoUpdateRoutine() {
+	// Wait until the object layer is ready
+	var objAPI ObjectLayer
+	for {
+		objAPI = newObjectLayerWithoutSafeModeFn()
+		if objAPI == nil {
+			time.Sleep(time.Second)
+			continue
+		}
+		break
+	}
+
+	ctx := context.Background()
+
+	switch v := objAPI.(type) {
+	case *xlZones:
+		runDataUsageInfoForXLZones(ctx, v, GlobalServiceDoneCh)
+	case *FSObjects:
+		runDataUsageInfoForFS(ctx, v, GlobalServiceDoneCh)
+	default:
+		return
+	}
+}
+
+func runDataUsageInfoForFS(ctx context.Context, fsObj *FSObjects, endCh chan struct{}) {
+	t := time.NewTicker(dataUsageCrawlInterval)
+	for {
+		// Get data usage info of the FS Object
+		usageInfo := fsObj.crawlAndGetDataUsageInfo(ctx, endCh)
+		// Save the data usage in the disk
+		err := storeDataUsageInBackend(ctx, fsObj, usageInfo)
+		if err != nil {
+			logger.LogIf(ctx, err)
+		}
+		select {
+		case <-endCh:
+			return
+		// Wait until the next crawl interval
+		case <-t.C:
+		}
+	}
+}
+
+func runDataUsageInfoForXLZones(ctx context.Context, z *xlZones, endCh chan struct{}) {
```

Comment:
> since endCh is globalDoneCh which is a read channel, Can we change the signature to `endCh <-chan struct{}`

GlobalServiceDoneCh is chan struct{} not <-chan struct{}

Reply:
I meant this to avoid writes to the GlobalServiceDoneCh not even by Mistake.. Like how all the NewTarget functions have (Eg, NewWebhookTarget). We need `<-chan struct{}` just to prevent any writes in case.